### PR TITLE
Add barcode detection feature

### PR DESCRIPTION
**Pull request by ralucatudor**
Using [ZXing](https://opensource.google/projects/zxing) (**barcode image processing library** implemented in Java), integrate barcode detection feature. 

Also, ZXing is the basis of Android’s Barcode Scanner app and is integrated into Google Product and Book Search :)

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/HandlePhotoShoppingServlet.java [hunk 1]
```diff
@@ -142,7 +143,14 @@ private String getQuery(String photoCategory, byte[] uploadedImageBytes)
...
       case "barcode":
-        return "Running shoes";
+        BarcodeImageDetector barcodeImageDetector = new BarcodeImageDetector();
+        String barcodeQuery;
+        try {
+          barcodeQuery = barcodeImageDetector.detect(uploadedImageBytes);
+        } catch (PhotoDetectionException exception) {
+          throw exception;
+        }
+        return barcodeQuery;
```
Comment (skopecki) [inline src/main/java/com/google/sps/servlets/HandlePhotoShoppingServlet.java:153]: Is there a benefit of catching an exception and then just throwing it again?
Wouldn't `return barcodeImageDetector.detect(uploadedImageBytes);` do the same?

File: src/main/java/com/google/sps/BarcodeImageDetector.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    // with black data on white backgrounds.
+    BinaryBitmap bitmap = new BinaryBitmap(new HybridBinarizer(
+        new BufferedImageLuminanceSource(bufferedImage)));
+    
+    // If needed, rotate the image.
+    if (bitmap.getWidth() < bitmap.getHeight()) {
+      if (bitmap.isRotateSupported()) {
+        bitmap = bitmap.rotateCounterClockwise();
+      }
+    }
```
Comment (skopecki) [inline src/main/java/com/google/sps/BarcodeImageDetector.java:75]: Does the library not work if the barcode is rotated? does it work if the barcode is upside down?

In any case I think the code does more bad than good: What if the image is in portrait mode (default on phone) but the barcode is oriented correctly?

If there's an issue with wrongly oriented barcodes which cannot be resolved by the library, then prefer to rotate the image again after the API fails to load barcode from the original image.

File: src/main/java/com/google/sps/BarcodeImageDetector.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  @AutoValue
+  public static abstract class BarcodeInfo {
+    
+    public static BarcodeInfo create(String text, String format) {
+      return new AutoValue_BarcodeImageDetector_BarcodeInfo(text, format);
+    }
+
+    public abstract String getText();
+    public abstract String getFormat();
+  }
```
Comment (skopecki) [inline src/main/java/com/google/sps/BarcodeImageDetector.java:108]: Are you planning to use the `getFormat` in future for something? If you don't plan to use it in a follow-up PR just stick with a simple text string.

File: src/main/java/com/google/sps/servlets/HandlePhotoShoppingServlet.java (hunk 1 above)
Comment (ralucatudor) [reply src/main/java/com/google/sps/servlets/HandlePhotoShoppingServlet.java:153]: Indeed, solved!

File: src/main/java/com/google/sps/BarcodeImageDetector.java (hunk 3 above)
Comment (ralucatudor) [reply src/main/java/com/google/sps/BarcodeImageDetector.java:108]: I had my doubts about that format, too. Indeed, I guess it won't be used and I'll get rid of the class. 👍

File: src/main/java/com/google/sps/BarcodeImageDetector.java (hunk 2 above)
Comment (ralucatudor) [reply src/main/java/com/google/sps/BarcodeImageDetector.java:75]: The detection works if it is upside down, but does not work if it is rotated at 90 degrees, so the code is helpful only when the barcode image is horizontal. Thank you for pointing out the problem, I haven't thought about the possibility that the photo is portrait and the barcode still horizontally placed. I thought that users would normally try to only capture the barcode, and therefore, if the picture's portrait, the barcode would also be vertical. 

I'll update as suggested and only rotate if the API fails

File: src/main/java/com/google/sps/BarcodeImageDetector.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import java.awt.image.BufferedImage;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * For a barcode image, detects the code, using the Java Apache Camel 
+ * Barcode API based on the ZXing library, and returns it.
+ */
+public class BarcodeImageDetector {
```
Comment (ChangchengXiao) [inline src/main/java/com/google/sps/BarcodeImageDetector.java:43]: Please add a constructor for this class.

File: src/main/java/com/google/sps/BarcodeImageDetector.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    try {
+      barcodeText = decode(bitmap);
+    } catch (PhotoDetectionException e) {
+      // Try to rotate the image and try decoding again.
+      if (bitmap.getWidth() < bitmap.getHeight()) {
+        if (bitmap.isRotateSupported()) {
+          bitmap = bitmap.rotateCounterClockwise();
+        }
+      }
+      return decode(bitmap);
```
Comment (ChangchengXiao) [inline src/main/java/com/google/sps/BarcodeImageDetector.java:80]: This should be handled within "decode" I think. The current implementation has the following problems:

#1 it swallows the PDE exception
#2 it has different behaviors for a rotated image.

File: src/main/java/com/google/sps/BarcodeImageDetector.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    } catch (NotFoundException e) {
+      // No potential barcode is found.
+      throw new PhotoDetectionException("No potential barcode found.", e);
+    } catch (ChecksumException e) {
+      // A potential barcode is found but does not pass its checksum.
+      throw new PhotoDetectionException("Potential barcode does not pass its checksum.", e);
+    } catch (FormatException e) {
+      // A potential barcode is found but format is invalid.
+      throw new PhotoDetectionException("Invalid format for potential barcode.", e);
+    }
```
Comment (ChangchengXiao) [inline src/main/java/com/google/sps/BarcodeImageDetector.java:99]: Ideally, the "decode" library should handle the rotation case. If it doesn't it is fine for you to do it in this class, I think.

**Review by ChangchengXiao — APPROVED** [review]

File: src/main/java/com/google/sps/BarcodeImageDetector.java (hunk 2 above)
Comment (skopecki) [reply src/main/java/com/google/sps/BarcodeImageDetector.java]: SGTM

**Review by skopecki — APPROVED** [review]

**Pull request merged**